### ISSUE TITLE
Stable recursive transforms for high azimuthal band limits

### ISSUE THREAD

**Issue by CosmoMatt** [astro-informatics/s2fft#227]
@ElisR picking up our previous discussion here. 

### Problem
It seems that you would like to efficiently evaluate the inverse Wigner transforms for L = N = 64? Are you interested in pushing this to higher L or is 64 sufficient? Current underlying recursion runs into stability issues for azimuthal band limits (N) greater than ~8, depending on L. Not clear precisely why, and for many use-cases this is acceptable, so we put priority on other projects. In any case, good to pick this up now!

### Background
As you suggested, retrofitting a more stable Wigner recursion for high N use-cases is the way to go. I've already implemented a JAX version of the Risbo recursion ([code](https://github.com/astro-informatics/s2fft/blob/efeeada245867871008f788afe2f34da4946016f/s2fft/recursions/risbo_jax.py#L7), [paper](https://www.google.com/search?client=safari&rls=en&q=Fourier+transform+summation+of+Legendre+series+and+D-+function&ie=UTF-8&oe=UTF-8)). @jasonmcewen and I have used this recursion in previous work (e.g. [paper](https://arxiv.org/pdf/1110.6298)) we have used this and found it is far more stable as N increases. Combining this recursion with the Fourier decomposition of Wigner D-functions presented by Trapani & Navaza ([paper](https://pubmed.ncbi.nlm.nih.gov/16788266/)) we designed efficient FFT based implementations for the spin-spherical harmonic ([paper](https://arxiv.org/pdf/1110.6298) equations 9-14) and Wigner transforms ([paper](https://arxiv.org/pdf/1508.03101) section III).

### Solution (Small L <= 64)
For your exact use case L = N = 64 is low enough that it should be possible to precompute all the necessary Wigner d-functions, which will then be **much faster at run-time**. This can be done at high N in a stable manner by using the Risbo recursion and following equation 8 of this [paper](https://arxiv.org/pdf/1110.6298). What we'll need to do is:

- [ ] Switch out the recursion used to precompute the Wigner d-functions (this [section](https://github.com/astro-informatics/s2fft/blob/efeeada245867871008f788afe2f34da4946016f/s2fft/precompute_transforms/construct.py#L284) in the code)

This will make the precompute take somewhat longer, but the memory complexity should remain the same $\mathcal{O}(NL^3)$ and the speed at run-time should be unaffected. We can reduce the memory complexity to $\mathcal{O}(L^3)$, which would be really good for your use case, but this will reduce the speed a little at runtime. In any case, I'll implement both versions so you can choose. What we'll need to do is:

- [ ]  Create alternate precompute transform which stores only $\Delta^{\ell}_{mn}$ and evaluates the necessary Wigner d-functions at run-time through equation 8 of this [paper](https://arxiv.org/pdf/1110.6298), i.e.

<img width="400" alt="image" src="https://github.com/user-attachments/assets/50491087-0ac4-42fd-99f3-a4f9d37d2b6f">

### Solution (Large L > 64)
We already have the core recursion necessary to solve the problem. What we'll need to do is: 

- [ ] Wrap the Risbo recursion into a forward and inverse spin-spherical harmonic transform.
- [ ] Wrap the spin-spherical harmonic transforms into Risbo based Wigner transforms.

I'll also need to check that memory/time complexity plays well for gradient propagation.

**Comment by CosmoMatt** [comment]
What I'd suggest, is if I implement the precompute approach first (starting presently) this should act as a stop-gap whilst I subsequently work on implementing the more scalable recursive approach.

**Comment by jasonmcewen** [comment]
Thanks for the clear overview @CosmoMatt.  For reference, while the Fourier decomposition of Wigner D-functions was used in Trapani & Navaza it was known well before that.

**Comment by ElisR** [comment]
Hi @CosmoMatt - thanks for the summary and for pointing me towards some of the literature.

Upon first inspection I think this should be doable for me if I dedicate a few hours to it this week (though apologies if it spills into next week).

Regarding my use case, $L = N = 64$ is probably the minimum I need, but I would also benefit from $L = 128$.
I'm not actually differentiating through this transform, so gradients are not critical for me, though I am of course in favour of a solution that works for all. 
The reason I need these high angular resolutions is because I'm aligning some spherical signals when calculating some training metrics, and $180^{\circ} / 64 \approx 3^{\circ}$ is around the acceptable threshold for my current signals.

#### Memory Estimate

I am just catching up so I'll run through a back-of-the-napkin calculation of memory lower bounds for a naive precomputation of all $d_{mn}^{\ell}(\beta)$. Please correct if I'm missing some symmetries / making a mistake.

Suppose $d_{mn}^{\ell}(\beta)$ is stored as one dense matrix for all $\ell <= \ell_{\text{max}} = 128 = 2^7$, then we have $128 * (2 * 128)^2 = 2^23$ matrix elements for each $\beta$. Since $d^{\ell}(\beta)$ is real, we need one real double (i.e. $8 = 2^3$ bytes) for each matrix element. Hence each $\beta$ requires $2^23 \times 2^3 = 2^26$ bytes i.e. `64MiB` for each $\beta$. Adding another factor of $2^7$ if we include all $\beta$, then we get $2^33$ bytes, i.e. `8GiB` for the entire precomputation. 

For my hardware, this is actually acceptable!
I hence support the plan of just swapping out the recursions for Risbo and seeing if things behave nicely without changing the logic to store $\Delta_{m n }^{\ell}$ etc.

#### Tests

I still need to look at the codebase a bit more to see what tests need to be modified to fairly assess the $N > 8$ performance, but I'm guessing that I can just follow the pattern of comparing to SSHT. I will also test on the small example code that I posted in #209.

**Comment by CosmoMatt** [comment]
@ElisR no worries, yep those back of those envelope memory calulcations look reasonable. You can see the issue with scalability here, given the O(L^4) memory scaling!

I actually spent some time tinkering over the weekend and should be able to push a MVP for this switch very shortly. There will of course still be some work to ensure it works for your application.

**Comment by ElisR** [comment]
Yep, I'm lucky to be one factor of 2 away from silly numbers 😅 .

Okay, that's great to hear, sounds good!

**Comment by sethaxen** [comment]
Hi, I came across this problem as well, and since I didn't see any work on this for a while, I thought I'd investigate it some more.

### Current state

I believe currently s2fft handles the instability in the Price-McEwen recurrence for |N|>~6 by switching to Risbo recurrence (#230), which is exceptionally stable but does not use the GPU as efficiently as Price-McEwen.

The [paper](https://doi.org/10.1016/j.jcp.2024.113109) in Section 3.3 after Eq (24) notes:
> While computing this recursion for increasing $`m`$ is unstable, it can be computed stably for decreasing $`m`$ (as discussed in detail in Pre'zeau and Reinecke[60], although they consider recursion over $`n`$).
> ...
> The recursion needs to be initialized at $`d_{\ell n}^\ell(\beta)`$ so that we can recurse down in $`m`$ starting from $`m=\ell`$. For this particular relation, recursing along decreasing harmonic order $`m`$ is stable, whereas the increasing $`m`$ direction causes numerical errors to compound.

Practically the columns are computed by downward recurrence for $`m=\ell\ldots0`$; for negative $`m`$, a symmetry relation is used by negating $`n`$, which allows for using positive $`m`$ and again a downward recurrence. 

### Cause of instability

From my reading of the [Prezeau and Reinecke paper](https://doi.org/10.1088/0067-0049/190/2/267) (and adapting to the recurrence in $`m`$ used here), recurrence in either direction *only* stably produces the physical solution when
```math
\ell(\ell+1)+\frac{2mn\cos\beta-m^2-n^2}{\sin^2\!\beta}\geq0,
\qquad\text{[Prézeau--Reinecke, Eq. (11)]}
```
Solving for $`m`$, this "classical region" has the bound
```math
m_\pm=n\cos\beta\pm\sin\beta\sqrt{\ell(\ell+1)-n^2}
```
Recurrence will be stable if we recur from $`|m|=\ell`$ *towards* this region and will become unstable and diverge once it exits the other side, where a nonphysical solution dominates the recurrence. Note the dependence on both $`\beta`$ and $`n`$. It is not true therefore that downward recurrence is always stable; the further one gets from $`\beta=\pi/2`$ and $`n=0`$, the higher the error will be if one recurs in the wrong direction.

### Solution

As is currently done, one could piece together an entire column $`m`$ from two recurrences: one from $`m=\ell`$ downwards and one from $`m=-\ell`$ upwards (or using a symmetry relation to recur downwards) and piece together the two recurrences. The only difference is the location of the seam $`p`$. Instead of using $`p=0`$, use the midpoint of the classical region $p=\lfloor\frac{1}{2}(m_+ + m_-)\rfloor = \lfloor n\cos\beta\rfloor$. Then one is always guaranteed to be recurring in a stable direction.

For a NumPy implementation of Price-McEwen recurrence, this can be done with the same amount of work as the original recurrence. For JAX, however, to do this in a way that can be `vmap`ed across spin and is independent of the values of `beta`, we probably would need to recur from $`m=\ell \to -\ell`$ and then do the same in reverse, using a `jnp.where` to determine which is kept based on the seam. This would roughly double the number of recurrence steps, so the main question is whether the benefits from better GPU parallelization relative to Risbo outweigh the extra ~2x recurrence cost.

I locally have a modified version of s2fft implementing this feature and would be happy to open a PR if there's interest. 

### Implementation/Validation

https://gist.github.com/sethaxen/ecc1ed282e62626a915a60a91f89b574 contains a standalone implementation of s2fft's current Price-McEwen recurrence as described in the paper with a `repaired` flag to indicate whether to use the repaired version proposed here. The below animations for the $`\ell=32`$ and $`\ell=64`$ planes shows on the left the regions where each recurrence is stable as a function of $`\beta`$, while all other subplots show absolute error relative to Risbo recurrence. The 2nd and 3rd subplots show s2fft's numerical instability and our reproducer of the same recurrence, and you can see that the high error occurs precisely when the chosen direction of recurrence (always from $`|m|=\ell \to 0`$) moves away from the classical region. The 4th subplot shows that the proposed repair maintains low error across the plane for all values of beta. The vertical dashed lines indicate spins where if requested currently s2fft switches to Risbo recurrence.

https://github.com/user-attachments/assets/408fa8bd-a01b-4f5b-9097-068c9a03559d

https://github.com/user-attachments/assets/566912d7-45e5-4642-a1a2-18b920a37c75
